### PR TITLE
Add ability to customize Dealer registration

### PR DESCRIPTION
**Pull request by kitsuta**
There was a lot of hard-coded functionality that made customization through a plugin difficult or impossible. This mostly keeps current functionality while allowing it to be overridden.

### CODE REVIEW

File: uber/model_checks.py [hunk 1]
```diff
@@ -37,6 +37,30 @@ def has_email_address(account):
...
+def dealer_description(group):
+    if group.tables and not group.description:
+        return "Please provide a brief description of your business"
+
+
+@validation.Group
+def dealer_address(group):
+    if group.tables and not group.address and not c.COLLECT_FULL_ADDRESS:
+        "Please provide your full address for tax purposes"
+
```
Comment (binary1230) [inline uber/model_checks.py]: should this be ```and c.COLLECT_FULL_ADDRESS``` so this error only fires off when we are collecting full addresses?

Comment (binary1230) [inline uber/model_checks.py]: i.e. changing ```and not c.COLLECT_FULL_ADDRESS``` to ```and c.COLLECT_FULL_ADDRESS```

File: uber/model_checks.py [hunk 2]
```diff
@@ -120,7 +144,7 @@ def emergency_contact(attendee):
 @validation.Attendee
 @ignore_unassigned_and_placeholders
 def cellphone(attendee):
-    if attendee.cellphone and _invalid_phone_number(attendee.cellphone):
+    if attendee.cellphone and _invalid_phone_number(attendee.cellphone) and not c.COLLECT_FULL_ADDRESS:
         return 'Your cellphone number was not a valid 10-digit phone number'
```
Comment (binary1230) [inline uber/model_checks.py]: same as comment above with the polarity of c.COLLECT_FULL_ADDRESS

Comment (binary1230) [inline uber/model_checks.py]: are we tying cellphone collection to full address collection?

File: uber/model_checks.py (hunk 1 above)
Comment (kitsuta) [inline uber/model_checks.py]: No. This validation deals with the textbox "Address" field for dealers. The idea here is that it's odd to ask them for an address, and then... ask them for an address again further down.

File: uber/templates/groups/form.html [hunk 3]
```diff
@@ -72,7 +72,7 @@
     <label class="col-sm-2 control-label">Tables</label>
     <div class="col-sm-6">
         <select name="tables">
-            {% int_options 0 9 group.tables %}
+            {% options c.ADMIN_TABLE_OPTS group.tables %}
         </select>
```
Comment (binary1230) [inline uber/templates/groups/form.html]: nice

File: uber/model_checks.py [hunk 4]
```diff
@@ -37,6 +37,30 @@ def has_email_address(account):
 
 
 @validation.Group
```
Comment (kitsuta) [inline uber/model_checks.py]: This is the one change in functionality here - by moving these to the `@validation.Group` class, these are now run every time a group is saved, including in the admin form. Hopefully that won't be too inconvenient.

File: uber/model_checks.py (hunk 1 above)
Comment (binary1230) [inline uber/model_checks.py]: oh wait, I think I get it.  we're saying that if c.COLLECT_FULL_ADDRESS is true, then the /group/ doesn't need to have an address collected or validated since we'll have a valid address from the attendee.

File: uber/model_checks.py [hunk 5]
```diff
@@ -120,7 +144,7 @@ def emergency_contact(attendee):
 @validation.Attendee
 @ignore_unassigned_and_placeholders
 def cellphone(attendee):
-    if attendee.cellphone and _invalid_phone_number(attendee.cellphone):
+    if attendee.cellphone and _invalid_phone_number(attendee.cellphone) and not c.COLLECT_FULL_ADDRESS:
```
Comment (kitsuta) [inline uber/model_checks.py]: `_invalid_phone_number`, among other things, checks to see if an attendee's phone number matches US-style 10-digit numbers. It skips validating numbers if the `international` checkbox is ticked - however, when we collect the full address, we don't use that checkbox. So this just bypasses the check in those cases.

File: uber/model_checks.py (hunk 1 above)
Comment (binary1230) [inline uber/model_checks.py]: I guess in some future universe this might be better (though a pain to change) if we rename c.COLLECT_FULL_ADDRESS to like c.REQUIRE_FULL_ATTENDEE_ADDRESS.  outside the scope of this PR

**Comment by binary1230** [comment]
looks good +1

**Comment by earl7399** [comment]
+1

File: uber/config.py [hunk 6]
```diff
@@ -363,6 +363,8 @@ def _is_intstr(s):
 
 c.TABLE_PRICES = defaultdict(lambda: _config['table_prices']['default_price'],
                              {int(k): v for k, v in _config['table_prices'].items() if k != 'default_price'})
```
Comment (EliAndrewC) [inline uber/config.py]: I guess I need to check my ``pep8`` config, since I'd have expected it to flag this with a style violation.  You're supposed to put spaces around operators, i.e. ``c.MAX_TABLES + 1`` instead of ``c.MAX_TABLES+1``.  Not a huge deal, but I'll try to remember and find the time to see why ``pep8`` didn't catch that in Travis.

File: uber/model_checks.py (hunk 4 above)
Comment (EliAndrewC) [inline uber/model_checks.py]: This will almost certainly be a problem.  The issue is that admins will often create groups which get tables which aren't dealers per se.  The main use case is bands, who have groups created for them by admins but don't have these fields filled out.

To give you some idea of how often this happens, here's a query from last year's database:

```python
>>> session.query(Group).filter(Group.tables > 0, Group.wares == '').count()
50
```

I wouldn't be comfortable with this change until we make sure that the affected admins are okay with it.  In the case of the ``wares`` column I could see still requiring it, but some of the other fields like ``website`` probably need to not be required on the admin side.

File: uber/model_checks.py (hunk 1 above)
Comment (EliAndrewC) [inline uber/model_checks.py]: +1 on that rename, but I agree that it's outside the scope of this PR.

File: uber/model_checks.py (hunk 5 above)
Comment (EliAndrewC) [inline uber/model_checks.py]: Hmm, so does this mean that when we collect the full address, we just don't validate phone numbers?

I don't have a huge problem with that (since it doesn't affect MAGFest), but that's extremely counter-intuitive, and I'd want to leave an ALL-CAPS-WARNING in ``configspec.ini`` indicating that our current implementation stops validating phone numbers if attendee address collection is turned on.

File: uber/templates/base.html [hunk 7]
```diff
@@ -165,7 +165,10 @@
     {% block head_additional %}{% endblock %}
 </head>
 <body>
-    {% if c.PRE_CON %}<div class="loader"><a class="loader_link" href="../static_views/slow_load.html" target="_blank"></a></div>{% endif %}
+    {% if c.PRE_CON %}
+        <div class="loader"><a class="loader_link" href="../static_views/slow_load.html" target="_blank"></a></div>
+        <script src="https://maps.googleapis.com/maps/api/js?sensor=false&amp;libraries=places"></script>
+    {% endif %}
```
Comment (EliAndrewC) [inline uber/templates/base.html]: +1 on the style fix.

File: uber/templates/groups/form.html (hunk 3 above)
Comment (EliAndrewC) [inline uber/templates/groups/form.html]: Ha, +1 on not hard-coding this stuff.

Out of curiosity, I just did a ``grep int_options uber -r`` and it looks like there's only one other place where we hard-code our ``int_options`` these days, and that's where we have a hard-coded limit of 50 "slots" for a job.  Seems like it would be easy to make that configurable as well, but probably not worth the effort, and definitely out of scope for this PR.

File: uber/templates/groups/form.html [hunk 8]
```diff
@@ -196,13 +196,14 @@
             <textarea rows="3" cols="30" name="special_needs">{{ group.special_needs }}</textarea>
         </div>
     </div>
```
Comment (EliAndrewC) [inline uber/templates/groups/form.html]: This doesn't affect MAGFest directly, so I don't care that much, but I'd be tempted to insert a note on the Attendee address when the dealer form is what's being loaded indicating that the address they enter MUST be their business address for tax purposes.  I can imagine plenty of situations where someone would enter their personal address instead of their business address (like if they expect the badge might get mailed to them).

Alternatively, it would probably be even better to show this field but make it optional when ``c.COLLECT_FULL_ADDRESS`` is turned on, with a note that you ONLY need to fill this out if your business' address is different from your personal address.  Kind of like how all the good online stores take your shipping address and then only make you fill out the billing address if it's different from the shipping address.

There are probably a LOT of businesses which are Delaware LLCs, which might mean a Delaware address for tax purposes.  But if an event was mailing out badges, someone would really NEED to put their own personal address.  So letting them enter both if they need to would be better.

But like I said, that doesn't affect MAGFest, so I'm okay with it being broken in that case, and an event which actually needed this could add it in their event-specific plugin if they wanted.

File: uber/config.py (hunk 6 above)
Comment (EliAndrewC) [inline uber/config.py]: Out of curiosity, why this change?  A plugin can say ``c.MAX_TABLES = 123`` to modify this config just as easily as they can modify ``c.PREREG_TABLE_OPTS``.

Relatedly, if you're already fixing the admin table value, maybe make it configurable while you're at it?  Not a huge deal if you're too swamped and don't have time, but if you're making changes anyway...

File: uber/templates/preregistration/preregbase.html [hunk 9]
```diff
@@ -129,7 +129,12 @@
         var setKickinFromBadge = function (types, index) {
             var type = types.options[index];
             if (type.extra !== undefined && type.extra !== null && $.field('amount_extra')) {
-                $('input:radio[name=amount_extra][value='+ type.extra +']').prop('checked', true).trigger('change');
+                {% if c.DONATION_SLIDER_DISPLAY %}
+                    $('input:radio[name=amount_extra][value='+ type.extra +']').prop('checked', true).trigger('change');
+                {% else %}
+                    $.field('amount_extra').val(type.extra);
+                    $.field('amount_extra').trigger('change');
+                {% endif %}
```
Comment (EliAndrewC) [inline uber/templates/preregistration/preregbase.html]: Was this brought in from a different branch?  I vaguely recall this showing up in a different PR, maybe one of the ones on Thursday that I haven't have time to review in detail yet.

File: uber/templates/regform.html [hunk 10]
```diff
@@ -89,7 +91,7 @@
 
 {% if c.PAGE == 'confirm' %}
     <div class="form-group">
-        <label for="email" class="col-sm-2 control-label">Legal Name</label>
+        <label for="full_name" class="col-sm-2 control-label">Legal Name</label>
```
Comment (EliAndrewC) [inline uber/templates/regform.html]: Good catch.  To be honest I've been very lax about the ``for`` attribute; I don't think I've even bothered to include it in any of the plugins I've written.  Which I really should.  Ah well, good fix.

File: uber/config.py (hunk 6 above)
Comment (kitsuta) [inline uber/config.py]: The real use-case here is when an event wants to define a set of tables that aren't integers. In Anthrocon's case, there are several specialized tables, including a half-table. You can see where I define them at: https://github.com/Anthrocon-Reg/anthrocon_plugin/blob/master/anthrocon_plugin/config.py#L6

File: uber/model_checks.py (hunk 4 above)
Comment (kitsuta) [inline uber/model_checks.py]: Daaangit.

On a related note, I really want a `@prereg_validation` class of some sort, so we don't even have to make this trade-off. I may look into seeing how easy that would be to do.

File: uber/model_checks.py (hunk 5 above)
Comment (kitsuta) [inline uber/model_checks.py]: Hm, okay. I'll think about a better way to do this.

File: uber/templates/groups/form.html (hunk 8 above)
Comment (kitsuta) [inline uber/templates/groups/form.html]: Yeah.. I think my preference here would be to have two sets of Amazon-style address fields, one of which is filled with the other if you check a "same as other address" checkbox. That's outside the scope of this for right now, though.

File: uber/templates/preregistration/preregbase.html (hunk 9 above)
Comment (kitsuta) [inline uber/templates/preregistration/preregbase.html]: Yup, I done goofed.

File: uber/templates/regform.html (hunk 10 above)
Comment (kitsuta) [inline uber/templates/regform.html]: `for=""` is an accessibility feature, so... we should definitely try to include it, yes.

File: uber/config.py (hunk 6 above)
Comment (EliAndrewC) [inline uber/config.py]: Ahh, that makes sense, since it also explains why you changed from ``int_options`` to ``options``.

Story checks out!

File: uber/model_checks.py (hunk 4 above)
Comment (EliAndrewC) [inline uber/model_checks.py]: Yeah, that was something I think I mentioned as a TODO item in the original PR that implemented ``@validation``.

The simplest approach (not necessarily the best) would be to define a second instance of the ``Validation`` class at
https://github.com/magfest/ubersystem/blob/master/uber/decorators.py#L316
like
```python
validation, prereg_validation = Validation(), Validation()
```

Then get rid of the ``validators`` property at
https://github.com/magfest/ubersystem/blob/master/uber/models.py#L232

which is only used in the ``check`` method anyway.  Instead of calling ``model.validators`` we could rewrite ``check``, which is at
https://github.com/magfest/ubersystem/blob/master/uber/utils.py#L63
to directly invoke the two different validations, taking an additional parameter to say whether or not it's a prereg validation.  It would look something vaguely like

```python
def check(model, *, prereg=False):
    # this first part is just exactly the same as what we have now
    for field, name in model.required:
        if not str(getattr(model, field)).strip():
            return name + ' is a required field'

    # this is basically just what we have now, but it checks both sets of validators if they say ``prereg=True``
    for v in [validation] + ([prereg_validation] if prereg else []):
        for validator in v[self.__class__.__name__].values()
            message = validator(model)
            if message:
                return message
```

If you did that, you could say ``@prereg_validation.Group`` and have it only run the validation if someone said ``check(group, prereg=True)``, which is presumably what you want.

I wanted to originally implement that, but time constraints, other excuses, etc.

File: uber/templates/regform.html (hunk 10 above)
Comment (EliAndrewC) [inline uber/templates/regform.html]: It's also useful even for non-accessibility reasons.  Like if you click on the label with the appropriate ``for`` label, it gives the input field focus.  That's helpful if someone wants to click on a field but isn't super-precise with the mouse.

**Comment by kitsuta** [comment]
OK, finally got around to updating this.
